src/coord/rmr/redise.c

```diff
@@ -20,16 +20,32 @@ typedef struct {
 static void MRTopology_AddRLShard(MRClusterTopology *t, RLShard *sh) {
 
   for (int i = 0; i < t->numShards; i++) {
-    if (sh->startSlot == t->shards[i].startSlot && sh->endSlot == t->shards[i].endSlot) {
-      // New node in the same shard (slot range)
-      MRClusterShard_AddNode(&t->shards[i], &sh->node);
-      return;
+    for (int r = 0; r < t->shards[i].numRanges; r++) {
+      if (sh->startSlot == t->shards[i].ranges[r].start && sh->endSlot == t->shards[i].ranges[r].end) {
+        // New node in the same shard
+        MRClusterShard_AddNode(&t->shards[i], &sh->node);
+        return;
+      }
+    }
+    if (!(sh->node.flags & MRNode_Master)) {
+      continue; // Only master nodes can cause shard merging
+    }
+    // Check if the master node already exists in the shard, in which case we can just
+    for (int n = 0; n < t->shards[i].numNodes; n++) {
+      if (!strcmp(sh->node.id, t->shards[i].nodes[n].id)) {
+        // Same node, extend the slot ranges (for all the nodes in the shard)
+        MRClusterShard_AddRange(&t->shards[i], sh->startSlot, sh->endSlot);
+        // Consume the node, we don't need to add it again
+        MRClusterNode_Free(&sh->node);
+        return;
+      }
     }
   }
 
   // New shard
-  MRClusterShard csh = MR_NewClusterShard(sh->startSlot, sh->endSlot, 2);
+  MRClusterShard csh = MR_NewClusterShard(2);
```

Comment:
I know it ws here, but can we give a meaning to this magic number with a #define?

Reply:
https://github.com/RediSearch/RediSearch/pull/6898#discussion_r2413459877
After the third PR is merged, we also won't need this capacity